src/sqs/index.ts

```diff
@@ -258,23 +311,67 @@ export class SqsQueue<MessagePayload> {
 			executionId: uuidv4(),
 			queue: this.queueName})
 
-		const context: Context<MessagePayload> = {message, payload, log }
+		const maybeReceivedCount: MessageAttributeValue | undefined = message.MessageAttributes ? message.MessageAttributes["ApproximateReceiveCount"] : undefined;
+
+		const receiveCount = maybeReceivedCount ? parseInt(maybeReceivedCount?.StringValue || "1") : 1
 
-		log.info("Sqs message received");
+		const context: Context<MessagePayload> = {message, payload, log, receiveCount: receiveCount}
+
+		log.info(`Sqs message received. Receive count: ${receiveCount}`);
 
 		try {
 			const messageProcessingStartTime = new Date().getTime();
-			await this.messageHandler.handle(context)
+
+			// Change message visibility timeout to the max processing time
+			// plus some room for error handling in case of timeout
+			await this.changeVisabilityTimeout(message, this.timeoutSec + 2, log);
+
+			const timeoutPromise = new Promise((_, reject) =>
+				setTimeout(() => reject(new SqsTimeoutError()), this.timeoutSec*1000)
+			);
+
+			await Promise.race([this.messageHandler.handle(context), timeoutPromise])
+
 			const messageProcessingDuration = new Date().getTime() - messageProcessingStartTime;
 			this.sendProcessedMetrics(messageProcessingDuration);
 			await this.deleteMessage(message, log)
 		} catch (err) {
+
+			const errorHandlingResult = this.errorHandler.handle(err, context);
+			if(!errorHandlingResult.retryable ||
+				(errorHandlingResult.skipDlq && context.receiveCount >= this.maxAttempts)) {
+				await this.deleteMessage(message, log)
+			} else {
+				if(errorHandlingResult.retryDelaySec) {
+					await this.changeVisabilityTimeout(message, errorHandlingResult.retryDelaySec, log);
+				}
+			}
+
 			//TODO Add error handling
 			statsd.increment(sqsQueueMetrics.failed, this.metricsTags)
 			log.error({err}, "error executing sqs message")
 		}
 	}
 
+	private async changeVisabilityTimeout(message: Message, timeout: number, logger: Logger): Promise<void> {
+
+		if(!message.ReceiptHandle) {
+			logger.error("NO Receipt Handle on a message")
```

Comment:
Could probably add the message to the data log to help debug

Reply:
The logger will already have messageId